services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/controller/SecurityGroupController.java

```diff
@@ -31,67 +34,129 @@
     @Autowired
     private SecurityGroupService securityGroupService;
 
+    private void checkProjectId(String projectId) throws ProjectIdRequired {
+        if (StringUtils.isEmpty(projectId)) {
+            throw new ProjectIdRequired();
+        }
+    }
+
+    private void checkTenantId(String tenantId) throws TenantIdRequired {
+        if (StringUtils.isEmpty(tenantId)) {
+            throw new TenantIdRequired();
+        }
+    }
+
+    private void checkSecurityGroup(SecurityGroupJson securityGroupJson) throws SecurityGroupRequired {
+        if (securityGroupJson == null || securityGroupJson.getSecurityGroup() == null) {
+            throw new SecurityGroupRequired();
+        }
+    }
+
+    private void checkSecurityGroupId(String securityGroupId) throws SecurityGroupIdRequired {
+        if (StringUtils.isEmpty(securityGroupId)) {
+            throw new SecurityGroupIdRequired();
+        }
+    }
+
+    private void checkPortId(String portId) throws PortIdRequired {
+        if (StringUtils.isEmpty(portId)) {
+            throw new PortIdRequired();
+        }
+    }
+
+    private void checkPortSecurityGroups(PortSecurityGroupsJson portSecurityGroupsJson) throws PortSecurityGroupsRequired {
+        if (portSecurityGroupsJson == null || portSecurityGroupsJson.getSecurityGroups() == null) {
+            throw new PortSecurityGroupsRequired();
+        }
+    }
+
+    private void checkSecurityGroups(SecurityGroupBulkJson securityGroupBulkJson) throws SecurityGroupsRequired {
+        if (securityGroupBulkJson == null ||
+                securityGroupBulkJson.getSecurityGroups() == null ||
+                securityGroupBulkJson.getSecurityGroups().size() == 0) {
+            throw new SecurityGroupsRequired();
+        }
+    }
+
     @PostMapping({"/project/{project_id}/security-groups", "v4/{project_id}/security-groups"})
     @ResponseBody
     @ResponseStatus(HttpStatus.CREATED)
     public SecurityGroupJson createSecurityGroup(@PathVariable("project_id") String projectId,
                                              @RequestBody SecurityGroupJson securityGroupJson) throws Exception {
-        RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectId);
-        RestPreconditionsUtil.verifyParameterNotNullorEmpty(securityGroupJson);
-        SecurityGroup securityGroup = securityGroupJson.getSecurityGroup();
-        RestPreconditionsUtil.verifyParameterNotNullorEmpty(securityGroup);
-        RestPreconditionsUtil.verifyParameterNotNullorEmpty(securityGroup.getTenantId());
-        securityGroup.setProjectId(projectId);
+        checkProjectId(projectId);
```

Comment:
In the future, we could write a general microservice validation framework, which supports an abstraction of validator and includes a list of abstract rules. A validator implementation for a specific microservice could include specific set of rules, for example, validateProjectId, validateSecurityGroup etc.